DeviceHarness/Program.cs

```diff
@@ -88,14 +94,67 @@ private static string GetTrainingSet()
             return setName;
         }
 
+        /// <summary>
+        /// Creates the set of tasks that will send data to the IoT hub.
+        /// </summary>
+        /// <param name="fileManager">TrainingFileManager for the data set for devices to send.</param>
+        /// <returns></returns>
+        private static List<Task> SetupDeviceRunTasks(TrainingFileManager fileManager)
+        {
+            List<Task> deviceTasks = new List<Task>();
+
+            string iotHubConnectionString = GetConnectionString();
+            int maxDevice = GetMaxDevice(fileManager);
+            string gatewayHost = gatewayHostNameOption.HasValue() ? gatewayHostNameOption.Value() : null;
+
+            for (int i = 1; i <= maxDevice; i++)
+            {
+                var device = new TurbofanDevice(i, iotHubConnectionString, fileManager, gatewayHost);
+                devices.Add(device);
+                deviceTasks.Add(device.RunDeviceAsync());
+            }
+
+            return deviceTasks;
+        }
+
+        /// <summary>
+        /// Return the number of devices for which to send, which will be the lower
+        /// of the number of devices in the data set and the value of the passed in max devices
+        /// </summary>
+        /// <param name="filemanager">TrainingFileManager for the data set for devices to send.</param>
+        /// <returns></returns>
+        private static int GetMaxDevice(TrainingFileManager filemanager)
+        {
+            int maxDevices = filemanager.MaxDeviceId;
+            int.TryParse(maxDevicesOption.Value(), out int maxRequested);
```

Comment:
if maxDevicesOption can be null, this will throw. You check on line 131 so you must think it could be null.